### PR TITLE
Non equality predicates in outer join (v2)

### PR DESCRIPTION
**Pull request by losipiuk**
Adds generic support for any conjuncts in join condition for outer joins.
Work on that implied changes in `JoinProbe` and `PagesHashStrategy`. This patch changes only non-generated implementations for those. If join condition contains non-equality conjuncts those are selected.
We will apply changes to code-compilers as soons as we agree on shape of changes needed.

@dain, @martint, @kokosing please review.

### CODE REVIEW

**Comment by losipiuk** [comment]
supersedes #4796

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 1]
```diff
@@ -264,7 +264,7 @@ else if (node.getType() == RIGHT) {
                 }
 
                 if (!isEqualComparisonExpression(conjunct) && node.getType() != INNER) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins only supported for inner join or must relate to inner side of outer join: %s", conjunct);
+                    throw new SemanticException(NOT_SUPPORTED, node, "Unsupported conjunct in join condition: %s", conjunct);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: you can squash this commit with the one before

File: presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java [hunk 2]
```diff
@@ -39,6 +41,7 @@
         private final SettableLookupSourceSupplier lookupSourceSupplier;
         private final List<Integer> hashChannels;
         private final Optional<Integer> hashChannel;
+        private final Optional<JoinFilterFunction> filterFunction;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java]: I would squash this commit with one before. As there you introduced an entity which is used here. When squashed, reader have a full story of this entity (JFF).

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 3]
```diff
@@ -1802,6 +1802,12 @@ public void testNonEqualityJoin()
         assertQuery("SELECT COUNT(*) FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey AND NOT (orders.comment LIKE lineitem.comment)");
         assertQuery("SELECT COUNT(*) FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey AND lineitem.quantity + length(orders.comment) > 7");
         assertQuery("SELECT COUNT(*) FROM lineitem JOIN orders ON lineitem.orderkey = orders.orderkey AND NULL");
+        assertQuery(
+                "SELECT * FROM (VALUES 1, 2) t1(a) JOIN (VALUES 10, 11) t2(b) ON a > 1",
+                "VALUES (2, 11), (2, 10)");
+        assertQuery(
+                "SELECT COUNT(*) FROM (VALUES 1, 2) t1(a) JOIN (VALUES 10, 11) t2(b) ON a > 2",
+                "VALUES (0)");
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: add OUTER JOINS with that predicate, as INNER can pushdown this predicate to table scan

what about `a < b` predicate?

Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: both cases are already covered

**Comment by losipiuk** [comment]
@martint, @dain please take a look when you have a chance.

**Comment by losipiuk** [comment]
@martint, @dain, @cberner Pinging - so this is not forgotten :)

File: presto-main/src/main/java/com/facebook/presto/operator/SimpleJoinProbe.java [hunk 4]
```diff
@@ -74,7 +74,7 @@ private SimpleJoinProbe(List<Type> types, LookupSource lookupSource, Page page,
             probeBlocks[i] = blocks[probeJoinChannels.get(i)];
         }
         this.page = page;
-        this.probePage = new Page(probeBlocks);
+        this.probePage = new Page(page.getPositionCount(), probeBlocks);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/SimpleJoinProbe.java]: Is this part of some other change?

File: presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java [hunk 5]
```diff
@@ -32,6 +32,7 @@
 public class SimplePagesHashStrategy
         implements PagesHashStrategy
 {
+    private static final Block[] EMPTY_BLOCK_ARRAY = new Block[0];
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java]: Is this part of some other change?

File: presto-main/src/main/java/com/facebook/presto/operator/JoinFilterFunction.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.Block;
+
+public interface JoinFilterFunction extends FilterFunction
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/JoinFilterFunction.java]: Place the `extends ...` on the next line.

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/JoinFilterFunction.java]: Should this `extend` `FilterFunction` at all?  It seem like this is a different concept than a `FilterFunction` so maybe they should just be separate interfaces.

**Comment by dain** [comment]
There is a typo in the `More generic support for non-equi conditions in outer joins` commit: conjucint

File: presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java (hunk 5 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/SimplePagesHashStrategy.java]: Removed. Some remainder from one of implementation approaches.

File: presto-main/src/main/java/com/facebook/presto/operator/JoinFilterFunction.java (hunk 6 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/JoinFilterFunction.java]: Yeah - probably. I will remove the inheritance on interface level and just make InterpretedFilterFunction implement two interfaces.

**Comment by losipiuk** [comment]
Applied comments in #5088

**Pull request closed without merging**